doc/manual/contract.md

```diff
@@ -0,0 +1,175 @@
+~~~
+
+* The `ensure` or `expect` contracts are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
+
+* In `ensure` contracts, the keyword `result` can only be used when the method has a return parameter.
+
+* The expression contained in an `old` is evaluated in the same context as the called method. Thus, it's possible to refer to attributes, methods, and arguments.
+
+* `old(object)` stores a reference to object. By doing so, if `object` is mutated in the method the `old (object)` will also be mutated. To avoid this it's necessary to provide a way to duplicate the object like `clone` method so that `not old (object.clone).is_same_instance (object)` will be true. In the stack example the expresion `old(stack).is_same_instance(stack) && old(stack.size) != old(stack).size` is true. If you need to `old(stack).is_same_instance(stack)` to be false, you need to provide a method to duplicate the object, such as `not old(stack.clone).is_same_instance(stack)` be true.
+
+* Contracts can be declared in several ways, all are equivalent:
+
+~~~
+	fun foo(a: Int, b: Int)
+	is
+		ensure(a > 0, b < 0)
+	do
+		# Do something
+	end
+
+
+	fun foo
+	is
+		ensure(a > 0 and b < 0)
+	do
+		# Do something
+	end
+
+
+	fun foo
+	is
+		ensure(a > 0)
+		ensure(b < 0)
+	do
+		# Do something
+	end
+~~~
+
+
+## Class/interface
+
+### Invariant
+
+Invariants are used to specify the characteristics of a class/interface which must always be true, except when executing a member function. In other words, invariants define conditions that must be respected by our instance. Any instance which does not respect one or more of those conditions will be considered as incoherent. Finally, invariants must hold before and after any method called (except those called by `self` directly).
+
+To define an invariant we use the class annotation `invariant`. Here is an example with our Stack. The class defines that it is impossible to be in a state where the size is less than 0.
+
+~~~
+	class Stack
+		invariant(size >= 0)
+	end
+~~~
+
+Another example with a class that represents a date (day, month). To be in a consistent state, the date must be represented with a day between 1 and 31 and a month between 1 and 12.
+
+~~~
+	class Date
+		invariant(day >= 1 and day <= 31)
+		invariant(month >= 1 and month <= 12)
+	end
+~~~
+
+### Additional informations
+
+* The invariant is only checked at the end of a constructor.
+
+* By default, the invariants are only checked on exit. See section `Verification policy` to activate them on in and out.
+
+## Inheritance
+
+The Nit contracts fully support inheritance (single or multiple) and refinement. The specialization can be interpreted as subcontracting. Subcontracting element must keep the set of promises initially defined by all previous contractors. Here are the subtyping rules for each type of the contracts:
```

Comment:
Is it the first time you use the `promises`? Why introducing a new concept that far in the document?

Reply:
Same for `contractors`